llvm/lib/Transforms/Utils/ScalarEvolutionExpander.cpp

```diff
@@ -1451,7 +1451,7 @@ Value *SCEVExpander::expandCodeFor(const SCEV *SH, Type *Ty) {
   // Expand the code for this SCEV.
   Value *V = expand(SH);
 
-  if (Ty) {
+  if (Ty && Ty != V->getType()) {
     assert(SE.getTypeSizeInBits(Ty) == SE.getTypeSizeInBits(SH->getType()) &&
```

Comment:
This is just a thought, but the assert also looks a bit odd here. Shouldn't it be

```
   assert(SE.getTypeSizeInBits(Ty) == SE.getTypeSizeInBits(V->getType()) &&
```

since we're casting between V->getType() and Ty? Is there something that guarantees SH->getType() == V->getType()?

Reply:
I *think* `expand` is guaranteed to return a value matching `SH`'s type. But in any case clarifying this would probably be best done separately, if needed.